From d004bd53a21c6569f4f9252a87810ffc8439721f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=81ngel=20Abreg=C3=BA?= <aj.abregu@yahoo.com.ar>
Date: Sun, 4 Nov 2012 02:39:46 -0300
Subject: [PATCH] Fix memory leaks (missing free's and pthread_detach)

---
 axel.c | 44 ++++++++++++++++++++++++++++++++++++++++----
 axel.h |  1 +
 2 files changed, 41 insertions(+), 4 deletions(-)

diff --git a/axel.c b/axel.c
index 84364cd..c7eddba 100644
--- a/axel.c
+++ b/axel.c
@@ -218,13 +218,17 @@ int axel_open( axel_t *axel )
 void axel_start( axel_t *axel )
 {
 	int i;
+	url_t* next_u = axel->url->next;
+	url_t* current_u = axel->url; 
 	
 	/* HTTP might've redirected and FTP handles wildcards, so
 	   re-scan the URL for every conn				*/
 	for( i = 0; i < axel->conf->num_connections; i ++ )
 	{
-		conn_set( &axel->conn[i], axel->url->text );
-		axel->url = axel->url->next;
+		conn_set( &axel->conn[i], current_u->text );
+		current_u = next_u;
+		next_u = next_u->next;
+		
 		axel->conn[i].local_if = axel->conf->interfaces->text;
 		axel->conf->interfaces = axel->conf->interfaces->next;
 		axel->conn[i].conf = axel->conf;
@@ -455,7 +459,10 @@ void axel_close( axel_t *axel )
 	for( i = 0; i < axel->conf->num_connections; i ++ )
 		/* don't try to kill non existing thread */
 		if ( *axel->conn[i].setup_thread != 0 )
+		{
 			pthread_cancel( *axel->conn[i].setup_thread );
+			pthread_detach( *axel->conn[i].setup_thread );
+		}
 	
 	/* Delete state file if necessary				*/
 	if( axel->ready == 1 )
@@ -482,8 +489,8 @@ void axel_close( axel_t *axel )
 	for( i = 0; i < axel->conf->num_connections; i ++ )
 		conn_disconnect( &axel->conn[i] );
 
-	free( axel->conn );
-	free( axel );
+	free_axel_t( axel );
+	free( buffer );
 }
 
 /* time() with more precision						*/
@@ -590,3 +597,32 @@ static void axel_divide( axel_t *axel )
 	printf( "Downloading %lld-%lld using conn. %i\n", axel->conn[i-1].currentbyte, axel->conn[i-1].lastbyte, i - 1 );
 #endif
 }
+
+
+
+void free_axel_t ( axel_t *axel )
+{
+	url_t *current_u = axel->url;
+	url_t *n_u = axel->url->next;
+	
+	while (  current_u != axel->url ) {
+		free( current_u );
+		current_u = n_u;
+		n_u =  n_u->next;
+	}
+	free( axel->url );
+	
+	
+	if_t *current_if = axel->conf->interfaces;
+	if_t *n_if = axel->conf->interfaces->next;
+	
+	while (  current_if != axel->conf->interfaces ) {
+		free( current_if );
+		current_if = n_if;
+		n_if = n_if->next;
+	}
+	free( axel->conf->interfaces );	
+	
+	free( axel->conn );
+	free( axel );
+}
diff --git a/axel.h b/axel.h
index e90d30d..ed755b3 100644
--- a/axel.h
+++ b/axel.h
@@ -112,3 +112,4 @@ void axel_do( axel_t *axel );
 void axel_close( axel_t *axel );
 
 double gettime();
+void free_axel_t ( axel_t *axel );
